### ISSUE TITLE
Routing Broken in Custom Controller / Loader

### ISSUE THREAD

**Issue by hparadiz** [simplesamlphp/simplesamlphp#1255]
I wrote my own loader for SimpleSAML because we need our framework to be loaded before SimpleSAML's code starts running for various reasons.

To make it work I manually set `$_SERVER['PATH_INFO']` and `$_SERVER['DOCUMENT_URI']`

This was all working perfectly until version 1.18

I'm going to go ahead and post the code to my custom controller below:
```php
class SimpleSAML extends RequestHandler
{
	public static function handleRequest()
	{
		static::bootstrap();

		$realpath = static::defineServerVariablesAndGetRealFile();

		if (!$realpath) {
			// throw 404
			die(404);
		} else {
			static::handleSimpleSAMLRequest($realpath);
		}
	}

	public static function bootstrap()
	{
		putenv(sprintf('SIMPLESAMLPHP_CONFIG_DIR=%s', App::$ApplicationPath.'/config/simplesaml'));
	}

	public static function getRoot()
	{
		return App::$ApplicationPath . '/vendor/simplesamlphp/simplesamlphp/www';
	}

	/**
	 * To handle SimpleSAML's weird router
	 * @return string Path to the file that is being routed to via /simplesaml/ url
	 */
	public static function defineServerVariablesAndGetRealFile()
	{
		$root = static::getRoot();

		// figure out the path and what file is being requested
		$path = parse_url($_SERVER['REQUEST_URI']);

		if (strpos($path['path'], '.php/') !== false) {
			$path = explode('.php/', $path['path'], 2); // limit 2 important!! cause of paths like /module.php/saml/sp/saml2-acs.php/onelogin-dev-local-sp
			$wanted = $path[0].'.php';
			$_SERVER['SCRIPT_FILENAME'] = $root.'/module.php';
		} else {
			$wanted = $path['path'];
		}
		$realpath = realpath($root.str_replace('/simplesaml', '', $wanted));

		$_SERVER['PATH_INFO'] = substr($_SERVER['REQUEST_URI'], strrpos($_SERVER['REQUEST_URI'], '/') + 1);
		$_SERVER['DOCUMENT_URI'] = $_SERVER['SCRIPT_NAME'] = $wanted;

		return $realpath;
	}

	public static function handleStaticFile($realpath)
	{
		if (substr($realpath, -4) === '.css') {
			$mimetype = 'text/css';
		} else {
			$mimetype = mime_content_type($realpath);
		}
		header('Content-Type: '.$mimetype);
		readfile($realpath);
		exit;
	}

	public static function handleSimpleSAMLRequest($realpath)
	{
		$root = static::getRoot();
		if ($realpath === $root) {
			require_once($realpath.'/index.php');
		} elseif (substr($realpath, -4) === '.php') { // other scripts
			chdir(dirname($realpath));
			require_once($realpath);
		} else { // assets
			static::handleStaticFile($realpath);
		}
	}
}
```

This might be related to the CRON controller issue as well.

Basically I just need to know what  `$_SERVER['PATH_INFO']` and ` $_SERVER['DOCUMENT_URI']` should be so I can shim in the proper result.

Thank you.

**Comment by hparadiz** [comment]
Just wanted to follow up with more information.

Here is my error.
<img width="932" alt="image" src="https://user-images.githubusercontent.com/195216/70325488-16372200-1800-11ea-9308-02d11116825d.png">


And here's the $_SERVER values when I breakpoint `Module::Process`

<img width="1868" alt="image" src="https://user-images.githubusercontent.com/195216/70325604-5f877180-1800-11ea-994d-85707e82e245.png">

@tvdijen @jaimeperez I would really appreciate your input. I see your commits to those files recently change how routing works.

**Comment by jaimeperez** [comment]
Hi @hparadiz,

I'm afraid I have to ask: what are those reasons to intercept SimpleSAMLphp before it starts execution? I'm asking because I don't see anything in your controller that actually should be there or justifies having code executed on beforehand.

In fact, you should not be messing around with `$_SERVER`, this is something that gets populated by your webserver and the PHP engine. If the issue in the first place was that `PATH_INFO` was not correctly populated (or not populated at all), you need to fix that in your deployment, but intercepting the execution to populate the value yourself and take over some of SimpleSAMLphp's responsibilities isn't the right way to do it.

**Comment by hparadiz** [comment]
@jaimeperez 

We need to use our own PDO connection to our database to load authsources and metadata. I made it use our own ORM to load that stuff and since our ORM wants to use our PDO connection it requires setting up autoloader for our framework before SimpleSAML even starts loading.

Among other framework specific reasons.

All I need is the values SimpleSAML expects for a valid request and I can emulate it no problem.

Basically if you could just tell me what the values for these 3 variables are for you when you breakpoint this line I can fix it.

<img width="1283" alt="image" src="https://user-images.githubusercontent.com/195216/70326585-e89fa800-1802-11ea-820c-1491a8b3df07.png">

**Comment by hparadiz** [comment]
I didn't really want to have to justify my reasons but this is important so I will explain in detail.

The current installation guide for SimpleSAML in the Nginx and Apache configuration guide says to use a config that more or less looks like this:
```nginx
server {
        listen 443 ssl;
        server_name idp.example.com;

        ssl_certificate        /etc/pki/tls/certs/idp.example.com.crt;
        ssl_certificate_key    /etc/pki/tls/private/idp.example.com.key;
        ssl_protocols          TLSv1.3 TLSv1.2;
        ssl_ciphers            EECDH+AESGCM:EDH+AESGCM;

        location ^~ /simplesaml {
            alias /var/www/vendor/simplesamlphp/simplesamlphp/www;

            location ~ ^(?<prefix>/simplesaml)(?<phpfile>.+?\.php)(?<pathinfo>/.*)?$ {
                include          fastcgi_params;
                fastcgi_pass     $fastcgi_pass;
                fastcgi_param    SCRIPT_FILENAME $document_root$fastcgi_script_name;
                fastcgi_param    SCRIPT_NAME /simplesaml$phpfile;
            }
        }
    }
```

There are obvious disadvantages to this.
- Changes require restarting the web server.
- The logic for rewriting $_SERVER['SCRIPT_FILENAME'] and $_SERVER['SCRIPT_NAME'] are stuck in a web server configuration file making refactoring your code reliant on a server.
- Logic changes in how nginx or apache interpret regular expressions and config variables can break SimpleSAML.

For example here's a [2014 blog post](https://casadelkrogh.dk/2014/09/embedding-simplesamlphp-using-nginx/) which goes into even more detail and makes several difficult to understand changes to the web server config.


This has downstream effects causing lots of confusion and backwards compatibility issues moving forward.

And unfortunately all this is for nothing. In fact it's not actually necessary to have custom logic in the web server config. This same logic can be entirely in the PHP code.

As an example here's the URL for the front page in SimpleSAML.
`/simplesaml/module.php/core/frontpage_welcome.php`.

All the information is right there. By default these variables are given to you as such:
<img width="444" alt="image" src="https://user-images.githubusercontent.com/195216/70330554-cfe7c000-180b-11ea-9cf0-be3af81c9dee.png">


Now... lets assume for a second that the config above is not being used. It is actually possible to have PHP do all the work. You can set $_SERVER['SCRIPT_FILENAME'] and $_SERVER['SCRIPT_NAME'] based on information already available to you at the very top and then run require() or include() on either index.php or module.php or whatever file you want inside the simplesaml www directory.

So that's what I've been doing.

**Comment by hparadiz** [comment]
I'm closing this cause I got it fixed on my end. Funny enough this was all I had to do
<img width="907" alt="image" src="https://user-images.githubusercontent.com/195216/70333039-b8f79c80-1810-11ea-8f41-0a9d5484396b.png">

**Comment by jaimeperez** [comment]
I'm glad you resolved your issue @hparadiz!

I'd like also to make some comments for the record, in case anybody is arriving to this conversation. In general, your concerns about configuring stuff in, well, web server **configuration**, are definitely not justified. Let me explain:

> We need to use our own PDO connection to our database to load authsources and metadata. I made it use our own ORM to load that stuff and since our ORM wants to use our PDO connection it requires setting up autoloader for our framework before SimpleSAML even starts loading.

Note that the configuration in SimpleSAMLphp is actual PHP code, so you can do whatever you need in there. If you need your own autoloader to load your framework so that you have everything you need to setup the PDO backend, then the proper way to do that is by adding an autoloader to your package/app with composer, not hardcoding it on your own by intercepting internal mechanisms of SimpleSAMLphp.

> * Changes require restarting the web server.

That's not true. I'm not familiar enough with nginx to make assertions about it, although I'm pretty sure it's the same case. In any case, with apache you don't need to restart the server when you change your configuration, you can tell it to reload it without restarting.

> * The logic for rewriting $_SERVER['SCRIPT_FILENAME'] and $_SERVER['SCRIPT_NAME'] are stuck in a web server configuration file making refactoring your code reliant on a server.

The problem here is not that you are modifying those global variables in the web server configuration, but the fact that you have to modify them. both `SCRIPT_FILENAME` and `SCRIPT_NAME` should have standard values very well defined, and you should never need to modify them. If your web server does not take care of that properly and you really need to do this manually, then the web server configuration is actually the right place where you should modify this, because that modification depends on your deployment. Modifying it in your own code is fragile because your modifications may not work properly if you change your deployment (e.g. switch to apache).

> * Logic changes in how nginx or apache interpret regular expressions and config variables can break SimpleSAML.

Here you are protecting against something that will never happen. Regular expressions are well defined, and a web server will simply not change how to _interpret regular expressions_. In fact there's nothing to interpret. Regular expressions have their own language and the meaning of all its elements is well established.

Regarding config variables, the issue is that we are not talking about config variables, but about standardized pieces of information. `PATH_INFO` is not up for interpretation either, it's [standardized](https://tools.ietf.org/html/rfc3875#section-4.1.5). That's also why you shouldn't need to ask software maintainers what values does their software expect for it, because those values depend on your deployment and are clearly defined. It's not up to SimpleSAMLphp to interpret anything here. The software assumes a correctly configured web environment. If your deployment does not provide the standard information it should, you need to fix your deployment (meaning your web server), not your software.

As a final remark: in order to avoid supposed backwards compatibility issues that will simply never happen, you are (ab)using internal software interfaces that are subject to change in future releases and could potentially break your setup.

**Comment by hparadiz** [comment]
With all due respect you are in fact incorrect about the vast majority of what you just wrote.

SimpleSAML is using an anti-pattern that was commonly in use 10 years ago and that is why the routing system is being refactored repeatedly. I'm trying to explain very patiently why it's a bad pattern but it looks like it's falling on deaf ears.

I have been working with PHP since 2003 so I hope you read my post again and try to understand the *why* more than the *how*.

I will go through your reply step by step.

> Note that the configuration in SimpleSAMLphp is actual PHP code, so you can do whatever you need in there. 

This is not a good solution. The config loads after a lot of application life-cycle events are loaded. Logging configuration, database configuration, and lots of other things need to be set up before SimpleSAML even loads. It is also the wrong place to put it and isn't namespaced or organized correctly. Not compliant with PSR specifications.

> That's not true. I'm not familiar enough with nginx to make assertions about it, although I'm pretty sure it's the same case. In any case, with apache you don't need to restart the server when you change your configuration, you can tell it to reload it without restarting.

Both Nginx and Apache rely on `kill -HUP {PID}` sent to the running process. They are both actually restart signals. The restart does happen instantly but it still *requires having terminal access to the server to make changes*.

Under the hood it is the same as running `sudo service nginx restart` or `sudo service apache restart`

> The problem here is not that you are modifying those global variables in the web server configuration, but the fact that you have to modify them. both `SCRIPT_FILENAME` and `SCRIPT_NAME` should have standard values very well defined, and you should never need to modify them.

Correct. Unfortunately your accepted solution for setting up SimpleSAML is what is modifying both SCRIPT_FILENAME and SCRIPT_NAME in favor of none standard values.

Here it is. You are telling people to add this to their server config because the default values break SimpleSAML.
```nginx
                fastcgi_param    SCRIPT_FILENAME $document_root$fastcgi_script_name;
                fastcgi_param    SCRIPT_NAME /simplesaml$phpfile;
```
It is reliant on this config for absolutely no reason. You could do everything you need to do without it.

> Regarding config variables, the issue is that we are not talking about config variables, but about standardized pieces of information. PATH_INFO is not up for interpretation either, it's standardized. 

That's exactly right but you guys are the ones modifying it.

Why are you messing with `SCRIPT_FILENAME` and `SCRIPT_NAME` at all if it's supposed to use the standard spec?

> Regarding config variables, the issue is that we are not talking about config variables, but about standardized pieces of information. PATH_INFO is not up for interpretation either, it's standardized. 

Yes. I am not actually messing with it. Your config expects SimpleSAML to be treated like a root path so if you visit `/simplesaml/index.php` the `PATH_INFO` variable would be `/index.php` but since my controller lives one directory up the value of `PATH_INFO` is `/simplesaml/index.php` so I'm cutting out the `/simplesaml` part of the string.

All I've done with my loader is emulate the same exact configuration that you are telling people to install...but with PHP code instead of web server configs.

> Here you are protecting against something that will never happen. Regular expressions are well defined, and a web server will simply not change how to interpret regular expressions. In fact there's nothing to interpret. Regular expressions have their own language and the meaning of all its elements is well established.

This also not correct. Both nginx and apache regularly fix issues with their regular expression engines and they are significantly different from PHP. Not all regular expression engines are the same. In fact there are huge difference just between PHP and JavaScript.  I have personally seen them break websites over the years because of changes in the logic of how mod_rewrite chooses to interpret something.

Just take a look at this [bug tracker for mod_rewrite](https://bz.apache.org/bugzilla/buglist.cgi?bug_status=__open__&component=mod_rewrite&list_id=144532&order=changeddate%20DESC%2Cpriority%2Cbug_severity&product=Apache%20httpd-2&query_format=specific). Any one of those being fixed could change how this line in your configuration is interpreted:
```location ~ ^(?<prefix>/simplesaml)(?<phpfile>.+?\.php)(?<pathinfo>/.*)?$ {```

My loader is likely to not break ever again unless you actually change the configuration.

I get why you don't want to make any changes to the current version but you should really consider not manually defining `SCRIPT_FILENAME` and `SCRIPT_NAME` in the next major version since that logic can be done entirely in SimpleSAML's router based only off of the standard `PATH_INFO` result.

The pattern I suggest you move to is documented here: https://laravel.com/docs/5.0/installation#pretty-urls

This is the same pattern that Wordpress, Laravel, and Symfony use and is considered "the right way" by the community.

**Comment by jaimeperez** [comment]
> SimpleSAML is using an anti-pattern that was commonly in use 10 years ago and that is why the routing system is being refactored repeatedly. I'm trying to explain very patiently why it's a bad pattern but it looks like it's falling on deaf ears.

Not very respectful for a reply that starts with "_with all due respect_", IMHO.

> I have been working with PHP since 2003 so I hope you read my post again and try to understand the why more than the how.

Appealing to a supposed authority based on how long you have been doing something is a logical fallacy, and you should avoid it. I have been working with PHP in one way or another since the end of last century. Do you see now why it's a mistake?

> This is not a good solution. The config loads after a lot of application life-cycle events are loaded. 

The configuration is just configuration. Of course, having logic in your configuration is something that can lead to issues and should be avoided in general, but there are lots of legitimate uses for that.

> Logging configuration, database configuration, and lots of other things need to be set up before SimpleSAML even loads.

That's simply not true. The configuration file is processed after SimpleSAMLphp has been _loaded_ (whatever that means, actually), and the configuration is **not** doing any kind of initialization. It's just assigning values to configuration options, period.

> It is also the wrong place to put it and isn't namespaced

I certainly hope you are not talking about PHP namespaces here, because that wouldn't make any sense at all.

> or organized correctly

That's absolutely true, and that's why [we are slowly trying to rationalize the configuration](https://github.com/simplesamlphp/simplesamlphp/issues/368). That doesn't invalidate the argument, though. If you need a specific configuration for your environment, the configuration is the place where you should do that.

> Not compliant with PSR specifications.

What does that even mean? Can you point me exactly to the PSR standard that defines how configuration  should be done? 

> Both Nginx and Apache rely on `kill -HUP {PID}` sent to the running process. They are both actually restart signals. The restart does happen instantly but it still requires having terminal access to the server to make changes.
> 
> Under the hood it is the same as running sudo service nginx restart or sudo service apache restart

Funny that you are telling me that I'm wrong and then saying something absolutely false like this. It might be the case for nginx, but it's definitely not the case for apache. [A _reload_ event in apache is equivalent to a **graceful** restart, which is using the `USR1` signal, not `HUP`](https://httpd.apache.org/docs/2.4/stopping.html#graceful).

> Correct. Unfortunately your accepted solution for setting up SimpleSAML is what is modifying both SCRIPT_FILENAME and SCRIPT_NAME in favor of none standard values.

Can you point me exactly to the point in the documentation are we saying that you **always** need to modify those variables? You can't, because [we never say so](https://simplesamlphp.org/docs/stable/simplesamlphp-install#section_4). Oh, and no, SimpleSAMLphp does not rely on non-standard values. In fact, the problem here is that your web server is not producing standard values.

 > Here it is. You are telling people to add this to their server config because the default values break SimpleSAML.

Wrong. We are telling people that they might need to configure nginx like that depending on their deployment. Pay attention to what you read in the documentation: _The configuration **may look like this**:_. There is a very good reason for that. Our experience shows that in some deployments, nginx will not provide proper values for those variables, hence the need to add those hints in the configuration to help those who might experience trouble with nginx.

In fact, [the nginx documentation itself](https://www.nginx.com/resources/wiki/start/topics/examples/phpfcgi/) is giving you configuration examples where it tells you that you might need to adjust those variables yourself. Surprisingly enough, all the examples there, from the default file in debian and ubuntu, to the custom configuration proposed, use **exactly the same configuration** that we are suggesting in the configuration:

```nginx
fastcgi_param   SCRIPT_FILENAME         $document_root$fastcgi_script_name;
```

If you do some research, you'll see that almost every example out there is telling you to set that variable. Unsurprisingly, not setting it results in no PHP in SimpleSAMLphp executed at all. That means the default provided by nginx, whatever that is, is not suitable to work with PHP-FPM, because your script doesn't even get executed. So no, there's no way we can solve this in code. You need to make nginx propagate the right data for PHP-FPM to work.

Similarly, `$_SERVER['PATH_INFO']` does not get populated by default in many linux distributions. In those where it does, it does because it is defined in `/etc/nginx/fastcgi_params`, which again is configuration, not code:

```nginx
fastcgi_param   PATH_INFO               $fastcgi_path_info;
```

All of this means that no, the webserver (nginx in this case) does not propagate the information as it should by default. And while that might be doable in code rather than in configuration, it's obvious from what you see when googling a little bit that the common practice is to fix **this issue with nginx in nginx itself**.

> It is reliant on this config for absolutely no reason. You could do everything you need to do without it.

No, you can't. No SimpleSAMLphp code will ever be executed in many deployments if you don't define `SCRIPT_FILENAME` manually in the web server configuration. So if you can't even run code without that, I'd say that qualifies as a very good reason to add that configuration.

> That's exactly right but you guys are the ones modifying it.

SimpleSAMLphp does not modify anything. **You** need to modify it if your deployment doesn't propagate that information with the values it should, by default.

> Why are you messing with SCRIPT_FILENAME and SCRIPT_NAME at all if it's supposed to use the standard spec?

You should be asking nginx why doesn't it propagate `SCRIPT_FILENAME` properly in a case like this.

> Yes. I am not actually messing with it. Your config expects SimpleSAML to be treated like a root path so if you visit /simplesaml/index.php the PATH_INFO variable would be /index.php but since my controller lives one directory up the value of PATH_INFO is /simplesaml/index.php so I'm cutting out the /simplesaml part of the string.

Again, wrong. Really, really wrong, actually. SimpleSAMLphp expects `PATH_INFO` to be empty in that case. According to the standard, the `PATH_INFO` corresponding a request URI like `/simplesaml/index.php` should be empty. That is what SSP expects. What SSP expects in `PATH_INFO` is, as defined by the standard, what comes **after the script**. Read the spec:

> _URL-encoded, the PATH_INFO string forms the extra-path component of
   the Script-URI (see section 3.3) which follows the SCRIPT_NAME part
   of that path._

So for a URI like `/simplesaml/module.php/core/frontpage_welcome.php`, `PATH_INFO` **must** be `/core/frontpage_welcome.php`. If your web server does not populate that value, then your web server is wrong and needs to be configured properly.

> This also not correct. Both nginx and apache regularly fix issues with their regular expression engines and they are significantly different from PHP. Not all regular expression engines are the same. In fact there are huge difference just between PHP and JavaScript. I have personally seen them break websites over the years because of changes in the logic of how mod_rewrite chooses to interpret something.

There are different flavours for regular expressions. One of the most common, used in this case by both nginx and PHP is PCRE. As you can see, [both apache and nginx use `libpcre`, which is also what PHP uses](https://en.wikipedia.org/wiki/Perl_Compatible_Regular_Expressions). So no, there's no way they are "_significantly different_", given that they use the same underlying library. The fact that there are differences between regexes in javascript and PHP is irrelevant here, as javascript uses a different flavour (xregexp).

Also, you are resorting to another argumentative fallacy here, the anecdotal fallacy. The fact that you have seen websites break because mod_rewrite changed something in the way it interpreted regular expressions doesn't make that an absolute truth. In fact, in more than ten years running SimpleSAMLphp in all kinds of setups, my experience says the contrary. I've never, ever, seen a web server break because a regular expression changed its meaning.

> Just take a look at this bug tracker for mod_rewrite. Any one of those being fixed could change how this line in your configuration is interpreted:
location ~ ^(?<prefix>/simplesaml)(?<phpfile>.+?\.php)(?<pathinfo>/.*)?$ {

Ehrm, no. I mean, theoretically they could. But then thousands of web servers would break, because that regex is absolutely basic.

> My loader is likely to not break ever again unless you actually change the configuration.

Your loader will break with SimpleSAMLphp 1.19 because you are using an experimental API that we will be changing. So you are achieving quite the contrary that you wanted.

> I get why you don't want to make any changes to the current version but you should really consider not manually defining SCRIPT_FILENAME and SCRIPT_NAME in the next major version since that logic can be done entirely in SimpleSAML's router based only off of the standard PATH_INFO result.

I'll say it one more time: **SimpleSAMLphp does not manually define any server variables, and it has never done**. Also, you can't do that logic based on `PATH_INFO` as you are suggesting if **`PATH_INFO` is missing or wrong**, as it happens to be the case for nginx in many setups.

> The pattern I suggest you move to is documented here: https://laravel.com/docs/5.0/installation#pretty-urls
> 
> This is the same pattern that Wordpress, Laravel, and Symfony use and is considered "the right way" by the community.

This is exactly what we are doing nowadays, and the reason why the APIs and interfaces are changing, as well as the router. It is also completely irrelevant to this discussion.
